From a6dc52cbcda9b4dcb0fda3b780e7c89219388982 Mon Sep 17 00:00:00 2001
From: Cristian Stoica <cristian.stoica@nxp.com>
Date: Mon, 8 Feb 2016 17:22:49 +0200
Subject: [PATCH 32/48] cryptodev: hide not used variable behind #ifndef

Signed-off-by: Cristian Stoica <cristian.stoica@nxp.com>
---
 crypto/engine/eng_cryptodev.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/crypto/engine/eng_cryptodev.c b/crypto/engine/eng_cryptodev.c
index b13bf8c..cdd99b8 100644
--- a/crypto/engine/eng_cryptodev.c
+++ b/crypto/engine/eng_cryptodev.c
@@ -413,7 +413,9 @@ static int open_dev_crypto(void)
 static int get_dev_crypto(void)
 {
     static int fd = -1;
+# ifndef CRIOGET_NOT_NEEDED
     int retfd;
+# endif
 
     if (fd == -1)
         fd = open_dev_crypto();
-- 
2.7.3

