From dcc3254b6dbb8627dd710fa58585542b98c80394 Mon Sep 17 00:00:00 2001
From: Cristian Stoica <cristian.stoica@nxp.com>
Date: Mon, 8 Feb 2016 17:11:43 +0200
Subject: [PATCH 38/48] cryptodev: match types with cryptodev.h

Signed-off-by: Cristian Stoica <cristian.stoica@nxp.com>
---
 crypto/engine/eng_cryptodev.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/crypto/engine/eng_cryptodev.c b/crypto/engine/eng_cryptodev.c
index 8cd3aa3..4613d2d 100644
--- a/crypto/engine/eng_cryptodev.c
+++ b/crypto/engine/eng_cryptodev.c
@@ -85,9 +85,9 @@ struct dev_crypto_state {
     int ivlen;
 # ifdef USE_CRYPTODEV_DIGESTS
     struct hash_op_data hash_op;
-    char dummy_mac_key[HASH_MAX_LEN];
+    unsigned char dummy_mac_key[HASH_MAX_LEN];
     unsigned char digest_res[HASH_MAX_LEN];
-    char *mac_data;
+    unsigned char *mac_data;
     int mac_len;
 # endif
 };
-- 
2.7.0

