From c79e7a4a818ea86bf6045197173d5c4e243d1f4f Mon Sep 17 00:00:00 2001
From: Cristian Stoica <cristian.stoica@nxp.com>
Date: Wed, 6 Apr 2016 15:22:58 +0300
Subject: [PATCH 48/48] fix 'maclen is used uninitialized' warning on some
 compilers

Signed-off-by: Cristian Stoica <cristian.stoica@nxp.com>
---
 crypto/engine/eng_cryptodev.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/crypto/engine/eng_cryptodev.c b/crypto/engine/eng_cryptodev.c
index ddd3462..2266b26 100644
--- a/crypto/engine/eng_cryptodev.c
+++ b/crypto/engine/eng_cryptodev.c
@@ -906,6 +906,10 @@ static int cryptodev_cbc_hmac_sha1_ctrl(EVP_CIPHER_CTX *ctx, int type,
                 maclen = SHA256_DIGEST_LENGTH;
                 aad_needs_fix = true;
                 break;
+            default:
+                fprintf(stderr, "%s: unsupported NID: %d\n",
+                        __func__, ctx->cipher->nid);
+                return -1;
             }
 
             /* Correct length for AAD Length field */
-- 
2.7.3

