From 970f04b228a60c805cce41ef8403ba2c8cb56d97 Mon Sep 17 00:00:00 2001
From: Otavio Salvador <otavio@ossystems.com.br>
Date: Wed, 13 Feb 2013 15:15:28 -0200
Subject: [PATCH] mx6q_sabresd: Change default environment to work with OE

Upstream-Status: Inappropriate [configuration]

Signed-off-by: Otavio Salvador <otavio@ossystems.com.br>
---
 include/configs/mx6q_sabresd.h | 40 ++++++++++++++++++++++++----------------
 1 file changed, 24 insertions(+), 16 deletions(-)

diff --git a/include/configs/mx6q_sabresd.h b/include/configs/mx6q_sabresd.h
index 6f07961..db31a34 100644
--- a/include/configs/mx6q_sabresd.h
+++ b/include/configs/mx6q_sabresd.h
@@ -116,22 +116,30 @@
 #define CONFIG_RD_LOADADDR	(0x1300000)
 
 #define	CONFIG_EXTRA_ENV_SETTINGS					\
-		"netdev=eth0\0"						\
-		"ethprime=FEC0\0"					\
-		"uboot=u-boot.bin\0"			\
-		"kernel=uImage\0"				\
-		"nfsroot=/opt/eldk/arm\0"				\
-		"bootargs_base=setenv bootargs console=ttymxc0,115200\0"\
-		"bootargs_nfs=setenv bootargs ${bootargs} root=/dev/nfs "\
-			"ip=dhcp nfsroot=${serverip}:${nfsroot},v3,tcp\0"\
-		"bootcmd_net=run bootargs_base bootargs_nfs; "		\
-			"tftpboot ${loadaddr} ${kernel}; bootm\0"	\
-		"bootargs_mmc=setenv bootargs ${bootargs} ip=dhcp "     \
-			"root=/dev/mmcblk0p1 rootwait\0"                \
-		"bootcmd_mmc=run bootargs_base bootargs_mmc; "   \
-		"mmc dev 3; "	\
-		"mmc read ${loadaddr} 0x800 0x2000; bootm\0"	\
-		"bootcmd=run bootcmd_net\0"                             \
+	 "script=boot.scr\0" \
+	 "uimage=uImage\0" \
+	 "mmcdev=2\0" \
+	 "mmcpart=1\0" \
+	 "mmcroot=/dev/mmcblk0p2 rw\0" \
+	 "mmcrootfstype=ext3 rootwait\0" \
+	 "mmcargs=setenv bootargs console=ttymxc3,${baudrate} " \
+	 "root=${mmcroot} " \
+	 "rootfstype=${mmcrootfstype}\0" \
+	 "loadbootscript=" \
+	 "fatload mmc ${mmcdev}:${mmcpart} ${loadaddr} ${script};\0" \
+	 "bootscript=echo Running bootscript from mmc ...; " \
+	 "source\0" \
+	 "loaduimage=fatload mmc ${mmcdev}:${mmcpart} ${loadaddr} ${uimage}\0" \
+	 "mmcboot=echo Booting from mmc ...; " \
+	 "run mmcargs; " \
+	 "bootm\0" \
+	 "netargs=setenv bootargs console=ttymxc3,${baudrate} " \
+	 "root=/dev/nfs " \
+	 "ip=dhcp nfsroot=${serverip}:${nfsroot},v3,tcp\0" \
+	 "netboot=echo Booting from net ...; " \
+	 "run netargs; " \
+	 "dhcp ${uimage}; bootm\0" \
+	 "bootcmd=run loaduimage; run mmcboot;\0" \
 
 
 #define CONFIG_ARP_TIMEOUT	200UL
-- 
1.8.1

