From 5066b8f46b0f2348881923491b93825ac04d13d2 Mon Sep 17 00:00:00 2001
From: Chunrong Guo <chunrong.guo@nxp.com>
Date: Sat, 29 Sep 2018 14:12:34 +0800
Subject: [PATCH] Makefile: fix gcc-8 build error

Signed-off-by: BJ DevOps Team <bjdevops@NXP1.onmicrosoft.com>
---
 Makefile | 1 +
 1 file changed, 1 insertion(+)

diff --git a/Makefile b/Makefile
index 53d57ac..b7c5fd4 100644
--- a/Makefile
+++ b/Makefile
@@ -395,6 +395,7 @@ LINUXINCLUDE	+= $(filter-out $(LINUXINCLUDE),$(USERINCLUDE))
 KBUILD_AFLAGS   := -D__ASSEMBLY__
 KBUILD_CFLAGS   := -Wall -Wundef -Wstrict-prototypes -Wno-trigraphs \
 		   -fno-strict-aliasing -fno-common \
+                   -Wno-packed-not-aligned -Wno-tautological-compare \
 		   -Werror-implicit-function-declaration \
 		   -Wno-format-security \
 		   -std=gnu89
-- 
1.8.3.1

