Index: lxc-0.8.0-rc1/src/lxc/Makefile.am
===================================================================
--- lxc-0.8.0-rc1.orig/src/lxc/Makefile.am
+++ lxc-0.8.0-rc1/src/lxc/Makefile.am
@@ -133,6 +133,7 @@ uninstall-local:
 	$(RM) $(DESTDIR)$(libdir)/liblxc.so*
 
 namespace.c: setns.h
+attach.c: setns.h
 
 setns.h:
 	-$(CC) $(CPPFLAGS) -M -MT$@ $(LINUX_DIR)/arch/$(LINUX_SRCARCH)/include/asm/unistd.h >setns.P
