From 9327cc35bc9e7040d9de743caebd0fee72bbd619 Mon Sep 17 00:00:00 2001
From: Chunrong Guo <chunrong.guo@nxp.com>
Date: Fri, 12 Oct 2018 10:53:00 +0800
Subject: [PATCH] automake: remove static key word

Upstream-Status: Inappropriate [embedded specific]

Signed-off-by: Chunrong Guo <chunrong.guo@nxp.com>
---
 lib/automake.mk | 1 +
 1 file changed, 1 insertion(+)

diff --git a/lib/automake.mk b/lib/automake.mk
index 8ecad12..f9016a0 100644
--- a/lib/automake.mk
+++ b/lib/automake.mk
@@ -439,6 +439,7 @@ lib/dhparams.c: lib/dh1024.pem lib/dh2048.pem lib/dh4096.pem
 	 openssl dhparam -C -in $(srcdir)/lib/dh2048.pem -noout &&	\
 	 openssl dhparam -C -in $(srcdir)/lib/dh4096.pem -noout)	\
 	| sed 's/\(get_dh[0-9]*\)()/\1(void)/' > lib/dhparams.c.tmp &&  \
+        sed -i '/\(get_dh[0-9]*\)(void)/s/^static//' lib/dhparams.c.tmp && \
 	mv lib/dhparams.c.tmp lib/dhparams.c
 else
 lib_libopenvswitch_la_SOURCES += lib/stream-nossl.c
-- 
2.7.4

