From ba3c9e624cf7ca370d6500c04428116939076bf1 Mon Sep 17 00:00:00 2001
From: Chunrong Guo <chunrong.guo@nxp.com>
Date: Wed, 10 Jul 2019 09:44:34 +0200
Subject: [PATCH] vpp-core:fix package_qa error

Upstream-Status: Inappropriate[oe-specific]

Signed-off-by: C.r. Guo <nxa13725@lsv07004.swis.us-cdc01.nxp.com>
---
 src/CMakeLists.txt | 1 -
 1 file changed, 1 deletion(-)

diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index 0326dbd..116bec3 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -51,7 +51,6 @@ endif()
 ##############################################################################
 # install config
 ##############################################################################
-set(CMAKE_INSTALL_RPATH "${CMAKE_INSTALL_PREFIX}/lib")
 set(CMAKE_INSTALL_MESSAGE NEVER)
 
 include_directories (
-- 
2.7.4

