diff --git a/mkspecs/linux-oe-g++/qmake.conf b/mkspecs/linux-oe-g++/qmake.conf
index c202c47fa1..2bfb96f4ec 100644
--- a/mkspecs/linux-oe-g++/qmake.conf
+++ b/mkspecs/linux-oe-g++/qmake.conf
@@ -33,6 +33,8 @@ QMAKE_CFLAGS_ISYSTEM =
 # for the SDK
 isEmpty(QMAKE_QT_CONFIG):QMAKE_QT_CONFIG = $$(OE_QMAKE_QT_CONFIG)
 
+EGLFS_DEVICE_INTEGRATION = eglfs_viv
+
 include(../oe-device-extra.pri)
 
 load(device_config)
-- 
2.17.1
