Index: git/mkspecs/linux-oe-g++/qmake.conf
===================================================================
--- git.orig/mkspecs/linux-oe-g++/qmake.conf	2016-12-14 17:03:17.000000000 -0600
+++ git/mkspecs/linux-oe-g++/qmake.conf	2016-12-14 17:06:23.000000000 -0600
@@ -39,6 +39,8 @@
 
 include(../oe-device-extra.pri)
 
+EGLFS_DEVICE_INTEGRATION = eglfs_viv
+
 QMAKE_LIBS_EGL        += -lEGL
 QMAKE_LIBS_OPENGL_ES2 += -lEGL -lGLESv2
 QMAKE_LIBS_OPENVG     += -lEGL -lOpenVG
