From fdb3add3dfd53c539c3c79e4fbf9eedb1dc1e440 Mon Sep 17 00:00:00 2001
From: Prabhu Sundararaj <prabhu.sundararaj@nxp.com>
Date: Fri, 26 Feb 2016 11:29:54 -0600
Subject: [PATCH] Use generic eglplatform.h rather than vendor specifc
 (eglvivante.h) EGL headers.

Upstream-Status: Pending

---
 wscript | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/wscript b/wscript
index 195e247..d08bc29 100644
--- a/wscript
+++ b/wscript
@@ -219,7 +219,7 @@ def check_vivante_egl(conf, egl_extra_defines):
 		extra_defines += egl_extra_defines
 	extra_cxxflags = map(lambda x: conf.env['DEFINES_ST'] % x, extra_defines)
 	conf.env['CXXFLAGS'] += extra_cxxflags
-	conf.check_cxx(mandatory = 1, header_name = 'EGL/eglvivante.h', uselib_store = 'EGL')
+	conf.check_cxx(mandatory = 1, header_name = 'EGL/eglplatform.h', uselib_store = 'EGL')
 	conf.check_cxx(mandatory = 1, header_name = 'EGL/egl.h', uselib_store = 'EGL')
 	conf.env['CXXFLAGS_EGL'] += extra_cxxflags
 	conf.env['CXXFLAGS'] = old_cxxflags
-- 
2.7.0

