diff --git a/EXA/src/makefile.linux b/EXA/src/makefile.linux
index 12cb8f3..005b002 100644
--- a/EXA/src/makefile.linux
+++ b/EXA/src/makefile.linux
@@ -1,9 +1,6 @@
 
 .PHONY: all clean install
 
-CC  := $(CROSS_COMPILE)gcc
-LD  := $(CROSS_COMPILE)gcc
-
 target := vivante_drv.so
 
 OBJS := vivante_gal/vivante_gal_blit.o \
@@ -63,7 +60,7 @@ LFLAGS += -L$(sysroot)/usr/lib -lm -ldl -lGAL -Wl,-soname -Wl,vivante_drv.so
 all: $(target)
 
 $(target): $(OBJS)
-	$(LD) -shared -fPIC -o $(target) $(OBJS) $(LFLAGS)
+	$(CC) -shared -fPIC -o $(target) $(OBJS) $(LFLAGS)
 
 clean:
 	-@rm $(target) $(OBJS)
