From 00a686189f7e05d70a7184cd6218f7424ab21b0d Mon Sep 17 00:00:00 2001
From: Cristian Stoica <cristian.stoica@nxp.com>
Date: Tue, 23 May 2017 15:28:58 +0300
Subject: [PATCH 5/9] remove unnecessary header inclusion

Signed-off-by: Cristian Stoica <cristian.stoica@nxp.com>
---
 zc.h | 2 --
 1 file changed, 2 deletions(-)

diff --git a/zc.h b/zc.h
index 6f975d6..666c4a5 100644
--- a/zc.h
+++ b/zc.h
@@ -1,8 +1,6 @@
 #ifndef ZC_H
 # define ZC_H
 
-#include "cryptodev_int.h"
-
 /* For zero copy */
 int __get_userbuf(uint8_t __user *addr, uint32_t len, int write,
 		unsigned int pgcount, struct page **pg, struct scatterlist *sg,
-- 
2.7.4

