From 787cbdf6b032964df00c4379d9a81ba4384a492e Mon Sep 17 00:00:00 2001
From: Cristian Stoica <cristian.stoica@nxp.com>
Date: Thu, 27 Oct 2016 10:06:08 +0300
Subject: [PATCH 073/104] really fix (again...) the mpstat issue

Signed-off-by: Cristian Stoica <cristian.stoica@nxp.com>
---
 tests/async_speed_multi.sh | 3 +--
 1 file changed, 1 insertion(+), 2 deletions(-)

diff --git a/tests/async_speed_multi.sh b/tests/async_speed_multi.sh
index e10dc0d..22a12cb 100755
--- a/tests/async_speed_multi.sh
+++ b/tests/async_speed_multi.sh
@@ -21,7 +21,6 @@
 
 NUM_CORES=$(nproc)
 OUT_BASENAME="async_speed"
-S_TIME_FORMAT=ISO
 MPSTAT_OUT="mpstat_out"
 
 function usage
@@ -62,7 +61,7 @@ function run_parallel
     echo "Running $tvalue seconds $mvalue threads in parallel:"
     echo "    $CMD"
 
-    (sleep 1; mpstat 1 $(($tvalue-2))) &> $MPSTAT_OUT &
+    (sleep 1; S_TIME_FORMAT=ISO mpstat 1 $(($tvalue-2))) &> $MPSTAT_OUT &
     MPSTAT_PID=$!
 
     PIDS=""
-- 
2.10.2

