From 6f75a9f4ad108c1f4a3dff4dd08def174ab83a69 Mon Sep 17 00:00:00 2001
From: Otavio Salvador <otavio@ossystems.com.br>
Date: Tue, 21 May 2013 16:43:31 -0300
Subject: [PATCH] configure.ac: Use pkg-config sysroot when checking for soname

Upstream-Status: Pending

Signed-off-by: Otavio Salvador <otavio@ossystems.com.br>
---
 configure.ac | 12 +++++-------
 1 file changed, 5 insertions(+), 7 deletions(-)

diff --git a/configure.ac b/configure.ac
index e4cdb86..c12ceb4 100755
--- a/configure.ac
+++ b/configure.ac
@@ -16,7 +16,7 @@ AC_DEFUN([FIND_PARSER_SONAME],
 [
     AC_MSG_CHECKING([for lib_$1_parser_$2_elinux.3.0.so soname])
     path=`$PKG_CONFIG --variable=libdir libfslparser`
-    SONAME=$( $CC -print-file-name=$path/lib_$1_parser_$2_elinux.so | \
+    SONAME=$( $CC -print-file-name=$PKG_CONFIG_SYSROOT_DIR$path/lib_$1_parser_$2_elinux.so | \
 	    while read output; do objdump -p $output | \
 	    grep "SONAME" | \
 	    sed -e 's/ \+SONAME \+//'; done 2> /dev/null )
@@ -27,16 +27,15 @@ AC_DEFUN([FIND_PARSER_SONAME],
     else
     	AC_MSG_RESULT([$SONAME])
     fi
-    apath=`echo $path | sed -e 's|.*/imx-mm/|/usr/lib/imx-mm/|'`
-    $1_$2_SONAME=$apath/$SONAME
+    $1_$2_SONAME=$path/$SONAME
     AC_SUBST($1_$2_SONAME)
 ])
 
 AC_DEFUN([FIND_ACODEC_SONAME],
 [
     AC_MSG_CHECKING([for lib_$1_wrap_$2_elinux.so soname])
-    path=`$PKG_CONFIG --variable=libdir libfslaudiocodec | sed -e 's|.*/imx-mm/|/usr/lib/imx-mm/|'`
-    SONAME=$( $CC -print-file-name=$path/wrap/lib_$1_wrap_$2_elinux.so | \
+    path=`$PKG_CONFIG --variable=libdir libfslaudiocodec`
+    SONAME=$( $CC -print-file-name=$PKG_CONFIG_SYSROOT_DIR$path/wrap/lib_$1_wrap_$2_elinux.so | \
 	    while read output; do objdump -p $output | \
 	    grep "SONAME" | \
 	    sed -e 's/ \+SONAME \+//'; done 2> /dev/null )
@@ -48,8 +47,7 @@ AC_DEFUN([FIND_ACODEC_SONAME],
     	AC_MSG_RESULT([$SONAME])
     fi
 
-    apath=`echo $path | sed -e 's|.*/imx-mm/|/usr/lib/imx-mm/|'`
-    $1_$2_SONAME=$apath/wrap/$SONAME
+    $1_$2_SONAME=$path/wrap/$SONAME
     AC_SUBST($1_$2_SONAME)
 ])
 
-- 
1.8.1

