diff --git a/arch/arm/plat-mxc/devices/platform-ahci-imx.c b/arch/arm/plat-mxc/devices/platform-ahci-imx.c
index c9267e1..069c93f 100644
--- a/arch/arm/plat-mxc/devices/platform-ahci-imx.c
+++ b/arch/arm/plat-mxc/devices/platform-ahci-imx.c
@@ -175,6 +175,6 @@ struct ahci_platform_data imx_sata_pdata = {
 #ifdef CONFIG_SOC_IMX53
 struct platform_device *__init imx53_add_ahci_imx(void)
 {
-	return imx_add_ahci_imx(&imx53_ahci_imx_data, &pdata);
+	return imx_add_ahci_imx(&imx53_ahci_imx_data, &imx_sata_pdata);
 }
 #endif
