diff --git a/drivers/mxc/Kconfig b/drivers/mxc/Kconfig
index c424f17..8b2a814 100644
--- a/drivers/mxc/Kconfig
+++ b/drivers/mxc/Kconfig
@@ -10,7 +10,7 @@ config MXC_IPU
 	depends on !SOC_MX27
 	depends on !SOC_MX25
 	select MXC_IPU_V1 if !SOC_MX37 && !SOC_MX5 && !SOC_IMX6Q
-	select MXC_IPU_V3 if SOC_MX37 || SOC_MX5 || SOC_IMX6Q
+	select MXC_IPU_V3 if SOC_MX37 || SOC_MX5 || SOC_IMX53 || SOC_IMX6Q
 	select MXC_IPU_V3D if SOC_MX37
 	select MXC_IPU_V3EX if SOC_MX5
 	select MXC_IPU_V3H if SOC_IMX6Q
