diff --git a/rcS b/rcS
index 080b04f..51425ff 100755
--- a/rcS
+++ b/rcS
@@ -22,6 +22,11 @@ export PATH runlevel prevlevel
 . /etc/default/rcS
 
 #
+#       Update the hostname according to actual board type
+#
+echo `cat /proc/cpuinfo | grep model | cut -d "," -f 2 | tr "[A-Z]" "[a-z]"` > /etc/hostname
+
+#
 #	Trap CTRL-C &c only in this shell so we can interrupt subprocesses.
 #
 trap ":" INT QUIT TSTP
