From 7070abd558e753b89955a966bb83e11f5a7b8462 Mon Sep 17 00:00:00 2001
From: Chris Dimich <Chris.Dimich@boundarydevices.com>
Date: Wed, 17 Feb 2021 14:06:26 -0800
Subject: [PATCH 1/1] run.sh: fix bashism

Signed-off-by: Chris Dimich <Chris.Dimich@boundarydevices.com>
---
 imx/run.sh | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/imx/run.sh b/imx/run.sh
index c0d676ae7..ebeb7a6e7 100755
--- a/imx/run.sh
+++ b/imx/run.sh
@@ -1,4 +1,4 @@
-#!/bin/bash
+#!/bin/sh
 #
 # Start the isp_media_server in the configuration from user
 # (c) NXP 2020
-- 
2.29.0

