From 5392152e704ff001c41de03445d668227237a87e Mon Sep 17 00:00:00 2001
From: Ting Liu <ting.liu@nxp.com>
Date: Fri, 15 Oct 2021 09:36:04 +0530
Subject: [PATCH] disable manpage generation

Upstream-Status: Inappropriate [Cross-compile specific]
Signed-off-by: Ting Liu <ting.liu@nxp.com>
---
 Makefile | 1 -
 1 file changed, 1 deletion(-)

diff --git a/Makefile b/Makefile
index ee669ae..ab6e499 100644
--- a/Makefile
+++ b/Makefile
@@ -60,7 +60,6 @@ install: restool scripts/ls-main scripts/ls-append-dpl scripts/ls-debug scripts/
 	install -D -m 755 scripts/ls-debug $(DESTDIR)$(bindir)/ls-debug
 	$(foreach symlink, $(RESTOOL_SCRIPT_SYMLINKS), sh -c "cd $(DESTDIR)$(bindir) && ln -sf ls-main $(symlink)" ;)
 	install -D -m 755 scripts/restool_completion.sh $(DESTDIR)$(bindir_completion)/restool
-	install -m 0644 -D $(MANPAGE) $(call get_manpage_destination,$(MANPAGE))
 
 clean:
 	rm -f $(OBJ) $(MANPAGE) \
-- 
2.25.1

