From 66c2285aff440d2cdbd16ee6893bdbbbd814a08b Mon Sep 17 00:00:00 2001
From: Carlos Rafael Giani <carlos.giani@streamunlimited.com>
Date: Tue, 7 Apr 2015 18:34:05 +0200
Subject: [PATCH] pxp: Check for new format specifier from the new FSL 3.14
 kernel

Signed-off-by: Carlos Rafael Giani <carlos.giani@streamunlimited.com>
---
 src/pxp/blitter.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/src/pxp/blitter.c b/src/pxp/blitter.c
index b96b3b9..7f44d03 100644
--- a/src/pxp/blitter.c
+++ b/src/pxp/blitter.c
@@ -391,7 +391,11 @@ static GstImxPxPFormatDetails const * gst_imx_pxp_blitter_get_pxp_format_details
 		case GST_VIDEO_FORMAT_YUY2: FORMAT_DETAILS(PXP_PIX_FMT_YUYV, 16);
 		case GST_VIDEO_FORMAT_UYVY: FORMAT_DETAILS(PXP_PIX_FMT_UYVY, 16);
 		case GST_VIDEO_FORMAT_YVYU: FORMAT_DETAILS(PXP_PIX_FMT_YVYU, 16);
+#ifdef PXP_PIX_FMT_VUY444
+		case GST_VIDEO_FORMAT_v308: FORMAT_DETAILS(PXP_PIX_FMT_VUY444, 24);
+#else
 		case GST_VIDEO_FORMAT_v308: FORMAT_DETAILS(PXP_PIX_FMT_YUV444, 24);
+#endif
 		case GST_VIDEO_FORMAT_IYU1: FORMAT_DETAILS(PXP_PIX_FMT_Y41P, 12);
 
 		/* planar formats; bits per pixel is always 8 for these */
-- 
1.9.1

