Index: chromium-96.0.4664.45/content/gpu/BUILD.gn
===================================================================
--- chromium-96.0.4664.45.orig/content/gpu/BUILD.gn
+++ chromium-96.0.4664.45/content/gpu/BUILD.gn
@@ -7,6 +7,11 @@ import("//build/config/ui.gni")
 import("//gpu/vulkan/features.gni")
 import("//media/media_options.gni")
 
+declare_args() {
+  # Checks if i.MX GPU is being used
+  use_imxgpu = false
+}
+
 # See //content/BUILD.gn for how this works.
 group("gpu") {
   visibility = [ "//content/*" ]  # This is an internal content API.
@@ -136,7 +141,7 @@ target(link_target_type, "gpu_sources")
 
   # Use DRI on desktop Linux builds.
   if (current_cpu != "s390x" && current_cpu != "ppc64" && is_linux &&
-      (!is_chromecast || is_cast_desktop_build)) {
+      (!is_chromecast || is_cast_desktop_build) && !use_imxgpu) {
     configs += [ "//build/config/linux/dri" ]
   }
 }
