From 895db76654c796a0edc6b2d329addde108b1755c Mon Sep 17 00:00:00 2001
From: Chunrong Guo <chunrong.guo@nxp.com>
Date: Wed, 19 Jul 2017 14:31:23 +0800
Subject: [PATCH] restool: fix build error with gcc7

---
 Makefile | 1 -
 1 file changed, 1 deletion(-)

diff --git a/Makefile b/Makefile
index 3976b66..df1e791 100644
--- a/Makefile
+++ b/Makefile
@@ -64,7 +64,6 @@ CFLAGS = ${EXTRA_CFLAGS} \
           -Wmissing-prototypes \
           -Wpointer-arith \
           -Winline \
-          -Werror \
           -Wundef \
           -Icommon/
 
-- 
2.7.4

