From 2fde2419acc8d89ec26c635a94c180899dbaa293 Mon Sep 17 00:00:00 2001
From: Chunrong Guo <chunrong.guo@nxp.com>
Date: Thu, 19 Sep 2019 05:27:11 +0200
Subject: [PATCH] Makfile: add modules_install

Signed-off-by: C.r. Guo <nxa13725@lsv07004.swis.us-cdc01.nxp.com>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index b67fad7..b386b39 100644
--- a/Makefile
+++ b/Makefile
@@ -191,7 +191,7 @@ clean:
 	@find $(AQROOT) -name ".gc_*.cmd" | xargs rm -f
 	@find $(AQROOT) -name "*.o" | xargs rm -f
 
-install:
+modules_install:
 	$(MAKE) -C $(KERNEL_DIR) M=$(PWD) modules_install
 
 else
-- 
2.7.4

